_posts/2025-02-05-#31363.md

```diff
@@ -10,17 +10,263 @@ status: upcoming
+approach is to first read through the overall changes to `TxGraph` and `TxGraphImpl`'s respective interfaces, and then
+review implementation. You could also review the interface, then the fuzzer implementation, then the real
+implementation.
+
+These notes are split into three sections by feature to make it easy to review a subset of the PR corresponding to the
+amount of time you have to review it.
+
+#### Basic TxGraph, Lazy and Oversized
 
+This section includes code up to [txgraph: (optimization) avoid per-group vectors for clusters & dependencies](https://github.com/bitcoin-core-review-club/bitcoin/commit/5e55a584a77a8dfa357eaf016f177d72cb92a218).
+
+A `TxGraph` has a vector of `Entry`s, where each entry is a single transaction, and a vector of `Cluster`s.  It also has
+queues of "work" to do which may or may not need to be completed in order to perform certain tasks or answer external
+queries. Some tasks have other tasks as prerequisites.
+
+This is the order in which things have to be resolved:
+
+1. `ApplyRemovals`: Apply all of the queued removals.
+2. `SplitAll`: Split up clusters if they contain components that are no longer connected to each other.
+3. `GroupClusters`: Calculate what clusters would need to be merged if dependencies are applied. However, these changes are not yet applied.
+4. `ApplyDependencies`: Apply new dependencies.
+5. Re-linearize clusters.
+
+Much of the work is done lazily. For example, calling `RemoveTransaction` does not actually mean its etry is deleted, just that
+the index has been added to `m_to_remove`. However, txgraph cannot answer a call to `Exists` without first calling `ApplyRemovals`.
+
+An **Oversized** `TxGraph` is one that has at least one cluster that violates the cluster count or size limits.
+- A `TxGraph` only knows if it is oversized after it has completed steps 1-3. 
+- When a `TxGraph` is oversized, mutation functions (like `RemoveTransaction` and `AddDependency`) and
+  `GetIndividualFeerate` can still be called, but inspection functions like `` are unavailable because we won't
+linearize clusters that are oversized.
+- Additionally, step 4 will not proceed if the txgraph is oversized; step 5 is thus also impossible.
+
+The `GroupClusters` function implements [union-find](https://en.wikipedia.org/wiki/Disjoint-set_data_structure), a
+favorite amongst software engineering interviewers.
+
+<!-- FIXME: Unlinked -->
+
+#### Staging Support
+
+This section includes code up to ["txgraph: (feature) add staging support"](https://github.com/bitcoin-core-review-club/bitcoin/commit/79647fd58dc7152eb1dc4e09cc56db845608352d).
+
+This feature allows users of `TxGraph` to stage changes and then apply or discard them, much like git.
+
+Much of the `TxGraph` members are grouped together as a `ClusterSet`: ther vector of `Cluster`s, work queues, and
+counter. This refactor is done so that a `TxGraph` can have multiple **subgraphs** or **levels** represented as
```

Comment:
s/ther/the, but I think the second part can just be removed - I think explaining the refactor is the main point here?
```suggestion
Much of the `TxGraph` members are grouped together as a `ClusterSet`. This refactor is done so that a `TxGraph` can have multiple **subgraphs** or **levels** represented as
```